// tgsnake/lib/Telegram/Messages/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnpinAllMessages = exports.sendMessage = exports.ReadMessageContents = exports.ReadMentions = exports.ReadHistory = exports.PinMessage = exports.GetMessagesViews = exports.Views = exports.ResultsMessagesViews = exports.ResultsGetMessage = exports.GetMessages = exports.ForwardMessages = exports.ExportMessageLink = exports.EditMessage = exports.DeleteUserHistory = exports.DeleteMessages = exports.ResultAffectedMessages = exports.DeleteHistory = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var DeleteHistory_1 = require("./DeleteHistory");
Object.defineProperty(exports, "DeleteHistory", { enumerable: true, get: function () { return DeleteHistory_1.DeleteHistory; } });
var DeleteMessages_1 = require("./DeleteMessages");
Object.defineProperty(exports, "ResultAffectedMessages", { enumerable: true, get: function () { return DeleteMessages_1.ResultAffectedMessages; } });
Object.defineProperty(exports, "DeleteMessages", { enumerable: true, get: function () { return DeleteMessages_1.DeleteMessages; } });
var DeleteUserHistory_1 = require("./DeleteUserHistory");
Object.defineProperty(exports, "DeleteUserHistory", { enumerable: true, get: function () { return DeleteUserHistory_1.DeleteUserHistory; } });
var EditMessage_1 = require("./EditMessage");
Object.defineProperty(exports, "EditMessage", { enumerable: true, get: function () { return EditMessage_1.EditMessage; } });
var ExportMessageLink_1 = require("./ExportMessageLink");
Object.defineProperty(exports, "ExportMessageLink", { enumerable: true, get: function () { return ExportMessageLink_1.ExportMessageLink; } });
var ForwardMessages_1 = require("./ForwardMessages");
Object.defineProperty(exports, "ForwardMessages", { enumerable: true, get: function () { return ForwardMessages_1.ForwardMessages; } });
var GetMessages_1 = require("./GetMessages");
Object.defineProperty(exports, "GetMessages", { enumerable: true, get: function () { return GetMessages_1.GetMessages; } });
Object.defineProperty(exports, "ResultsGetMessage", { enumerable: true, get: function () { return GetMessages_1.ResultsGetMessage; } });
var GetMessagesViews_1 = require("./GetMessagesViews");
Object.defineProperty(exports, "ResultsMessagesViews", { enumerable: true, get: function () { return GetMessagesViews_1.ResultsMessagesViews; } });
Object.defineProperty(exports, "Views", { enumerable: true, get: function () { return GetMessagesViews_1.Views; } });
Object.defineProperty(exports, "GetMessagesViews", { enumerable: true, get: function () { return GetMessagesViews_1.GetMessagesViews; } });
var PinMessage_1 = require("./PinMessage");
Object.defineProperty(exports, "PinMessage", { enumerable: true, get: function () { return PinMessage_1.PinMessage; } });
var ReadHistory_1 = require("./ReadHistory");
Object.defineProperty(exports, "ReadHistory", { enumerable: true, get: function () { return ReadHistory_1.ReadHistory; } });
var ReadMentions_1 = require("./ReadMentions");
Object.defineProperty(exports, "ReadMentions", { enumerable: true, get: function () { return ReadMentions_1.ReadMentions; } });
var ReadMessageContents_1 = require("./ReadMessageContents");
Object.defineProperty(exports, "ReadMessageContents", { enumerable: true, get: function () { return ReadMessageContents_1.ReadMessageContents; } });
var sendMessage_1 = require("./sendMessage");
Object.defineProperty(exports, "sendMessage", { enumerable: true, get: function () { return sendMessage_1.sendMessage; } });
var UnpinAllMessages_1 = require("./UnpinAllMessages");
Object.defineProperty(exports, "UnpinAllMessages", { enumerable: true, get: function () { return UnpinAllMessages_1.UnpinAllMessages; } });
