// tgsnake/lib/Context/Error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const CleanObject_1 = require("../Utils/CleanObject");
const util_1 = require("util");
class BotError extends Error {
    constructor(message, functionName, functionArgs) {
        super();
        this.date = Math.floor(Date.now() / 1000);
        /** @hidden */
        this._isBotErrorClass = true;
        this.message = message;
        this.functionName = functionName;
        this.functionArgs = functionArgs;
        return this;
    }
    /** @hidden */
    [util_1.inspect.custom]() {
        return (0, CleanObject_1.betterConsoleLog)(this);
    }
    /** @hidden */
    toJSON() {
        let obj = (0, CleanObject_1.betterConsoleLog)(this);
        for (let [key, value] of Object.entries(obj)) {
            if (typeof value == 'bigint')
                obj[key] = String(value);
        }
        return obj;
    }
}
exports.default = BotError;
