// tgsnake/lib/Context/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.EntityCache = exports.MainContext = exports.BotError = exports.MessageContext = exports.Composer = exports.run = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const Error_1 = __importDefault(require("./Error"));
exports.BotError = Error_1.default;
var Composer_1 = require("./Composer");
Object.defineProperty(exports, "run", { enumerable: true, get: function () { return Composer_1.run; } });
Object.defineProperty(exports, "Composer", { enumerable: true, get: function () { return Composer_1.Composer; } });
var MessageContext_1 = require("./MessageContext");
Object.defineProperty(exports, "MessageContext", { enumerable: true, get: function () { return MessageContext_1.MessageContext; } });
var MainContext_1 = require("./MainContext");
Object.defineProperty(exports, "MainContext", { enumerable: true, get: function () { return MainContext_1.MainContext; } });
var EntityCache_1 = require("./EntityCache");
Object.defineProperty(exports, "EntityCache", { enumerable: true, get: function () { return EntityCache_1.EntityCache; } });
var Logger_1 = require("./Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return Logger_1.Logger; } });
