// tgsnake/lib/Telegram/Media/EditPhoto.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditPhoto = void 0;
const telegram_1 = require("telegram");
const UploadFile_1 = require("./UploadFile");
const Updates = __importStar(require("../../Update"));
const ToBigInt_1 = require("../../Utils/ToBigInt");
const Error_1 = __importDefault(require("../../Context/Error"));
/**
 * Change the profile picture of chat.
 * @param snakeClient - Client
 * @param {number|string|bigint} chatId - Chat/Groups/Channel which will change the profile picture.
 * @param {string|Buffer} photo - The location where the file is located/Url/The buffer of the file.
 * ```ts
 * bot.command("editPhoto", async (ctx)=>{
 *     if(!ctx.chat.private){
 *         let results = await ctx.telegram.editPhoto(ctx.chat.id,"https://tgsnake.js.org/images/tgsnake.jpg")
 *         console.log(results)
 *     }
 * })
 * ```
 */
async function EditPhoto(snakeClient, chatId, photo) {
    try {
        snakeClient.log.debug('Running telegram.editPhoto');
        if (typeof chatId === 'number')
            snakeClient.log.warning('Type of chatId is number, please switch to BigInt or String for security Ids 64 bit int.');
        let rr = await (0, UploadFile_1.UploadFile)(snakeClient, photo);
        let [id, type, peer] = await (0, ToBigInt_1.toBigInt)(chatId, snakeClient);
        if (type == 'channel' || type == 'supergroup') {
            let results = await snakeClient.client.invoke(new telegram_1.Api.channels.EditPhoto({
                channel: id,
                photo: new telegram_1.Api.InputChatUploadedPhoto({
                    file: rr,
                }),
            }));
            return await generateResults(results, snakeClient);
        }
        else {
            return snakeClient.client.invoke(new telegram_1.Api.messages.EditChatPhoto({
                chatId: id,
                photo: new telegram_1.Api.InputChatUploadedPhoto({
                    file: rr,
                }),
            }));
        }
    }
    catch (error) {
        snakeClient.log.error('Failed to running telegram.editPhoto');
        throw new Error_1.default(error.message, 'telegram.editPhoto', `${chatId}${Buffer.isBuffer(photo) ? `<Buffer ${photo.toString('hex')} >` : photo}`);
    }
}
exports.EditPhoto = EditPhoto;
async function generateResults(results, SnakeClient) {
    SnakeClient.log.debug('Creating results telegram.editPhoto');
    if (results instanceof telegram_1.Api.Updates) {
        results;
        if (results.updates.length > 0) {
            for (let i = 0; i < results.updates.length; i++) {
                let update = results.updates[i];
                if (update instanceof telegram_1.Api.UpdateNewMessage) {
                    update;
                    let res = new Updates.UpdateNewMessage();
                    await res.init(update, SnakeClient);
                    return res;
                }
                if (update instanceof telegram_1.Api.UpdateNewChannelMessage) {
                    update;
                    let res = new Updates.UpdateNewChannelMessage();
                    await res.init(update, SnakeClient);
                    return res;
                }
            }
        }
    }
}
