// tgsnake/lib/Telegram/Media/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendInvoice = exports.SendVenue = exports.SendLocation = exports.SendPoll = exports.SendContact = exports.SendDice = exports.SendVoice = exports.SendAudio = exports.SendAnimation = exports.SendVideoNote = exports.SendVideo = exports.Download = exports.UploadFile = exports.inRange = exports.SendSticker = exports.SendMedia = exports.SendPhoto = exports.SendDocument = exports.GetFileInfo = exports.EditPhoto = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var EditPhoto_1 = require("./EditPhoto");
Object.defineProperty(exports, "EditPhoto", { enumerable: true, get: function () { return EditPhoto_1.EditPhoto; } });
var GetFileInfo_1 = require("./GetFileInfo");
Object.defineProperty(exports, "GetFileInfo", { enumerable: true, get: function () { return GetFileInfo_1.GetFileInfo; } });
var SendDocument_1 = require("./SendDocument");
Object.defineProperty(exports, "SendDocument", { enumerable: true, get: function () { return SendDocument_1.SendDocument; } });
var SendPhoto_1 = require("./SendPhoto");
Object.defineProperty(exports, "SendPhoto", { enumerable: true, get: function () { return SendPhoto_1.SendPhoto; } });
var SendMedia_1 = require("./SendMedia");
Object.defineProperty(exports, "SendMedia", { enumerable: true, get: function () { return SendMedia_1.SendMedia; } });
var SendSticker_1 = require("./SendSticker");
Object.defineProperty(exports, "SendSticker", { enumerable: true, get: function () { return SendSticker_1.SendSticker; } });
var UploadFile_1 = require("./UploadFile");
Object.defineProperty(exports, "inRange", { enumerable: true, get: function () { return UploadFile_1.inRange; } });
Object.defineProperty(exports, "UploadFile", { enumerable: true, get: function () { return UploadFile_1.UploadFile; } });
var Download_1 = require("./Download");
Object.defineProperty(exports, "Download", { enumerable: true, get: function () { return Download_1.Download; } });
var SendVideo_1 = require("./SendVideo");
Object.defineProperty(exports, "SendVideo", { enumerable: true, get: function () { return SendVideo_1.SendVideo; } });
var SendVideoNote_1 = require("./SendVideoNote");
Object.defineProperty(exports, "SendVideoNote", { enumerable: true, get: function () { return SendVideoNote_1.SendVideoNote; } });
var SendAnimation_1 = require("./SendAnimation");
Object.defineProperty(exports, "SendAnimation", { enumerable: true, get: function () { return SendAnimation_1.SendAnimation; } });
var SendAudio_1 = require("./SendAudio");
Object.defineProperty(exports, "SendAudio", { enumerable: true, get: function () { return SendAudio_1.SendAudio; } });
var SendVoice_1 = require("./SendVoice");
Object.defineProperty(exports, "SendVoice", { enumerable: true, get: function () { return SendVoice_1.SendVoice; } });
var SendDice_1 = require("./SendDice");
Object.defineProperty(exports, "SendDice", { enumerable: true, get: function () { return SendDice_1.SendDice; } });
var SendContact_1 = require("./SendContact");
Object.defineProperty(exports, "SendContact", { enumerable: true, get: function () { return SendContact_1.SendContact; } });
var SendPoll_1 = require("./SendPoll");
Object.defineProperty(exports, "SendPoll", { enumerable: true, get: function () { return SendPoll_1.SendPoll; } });
var SendLocation_1 = require("./SendLocation");
Object.defineProperty(exports, "SendLocation", { enumerable: true, get: function () { return SendLocation_1.SendLocation; } });
var SendVenue_1 = require("./SendVenue");
Object.defineProperty(exports, "SendVenue", { enumerable: true, get: function () { return SendVenue_1.SendVenue; } });
var SendInvoice_1 = require("./SendInvoice");
Object.defineProperty(exports, "SendInvoice", { enumerable: true, get: function () { return SendInvoice_1.SendInvoice; } });
