// tgsnake/lib/Telegram/Media/SendVenue.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendVenue = void 0;
const telegram_1 = require("telegram");
const SendMedia_1 = require("./SendMedia");
const Error_1 = __importDefault(require("../../Context/Error"));
/**
 * Sending Venue
 * @param snakeClient - Client
 * @param {string|number|bigint} chatId - Chat/Groups/Channel id.
 * @param {Object} venue - venue
 * @param {Object} more - more parameters to use.
 * ```ts
 * bot.command("venue",async (ctx) => {
 *     let results = await ctx.telegram.sendVenue(ctx.chat.id,{
 *    latitude : 0,
 *    longitude : 0,
 *    title : "title",
 *    address : "address",
 *    provider : "provider",
 *    id : "some id here",
 *    type : "some type here"
 *  })
 * })
 * ```
 */
async function SendVenue(snakeClient, chatId, venue, more) {
    try {
        snakeClient.log.debug('Running telegram.sendVenue');
        if (typeof chatId === 'number')
            snakeClient.log.warning('Type of chatId is number, please switch to BigInt or String for security Ids 64 bit int.');
        return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaVenue({
            geoPoint: new telegram_1.Api.InputGeoPoint({
                lat: venue.latitude,
                long: venue.longitude,
                accuracyRadius: venue.accuracyRadius,
            }),
            title: venue.title,
            address: venue.address,
            provider: venue.provider,
            venueId: venue.id,
            venueType: venue.type,
        }), more);
    }
    catch (error) {
        snakeClient.log.error('Failed to running telegram.sendVenue');
        throw new Error_1.default(error.message, 'telegram.sendVenue', `${chatId},${JSON.stringify(venue)}${more ? ',' + JSON.stringify(more) : ''}`);
    }
}
exports.SendVenue = SendVenue;
