// tgsnake/lib/Telegram/Media/SendVoice.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendVoice = void 0;
const telegram_1 = require("telegram");
const SendMedia_1 = require("./SendMedia");
const UploadFile_1 = require("./UploadFile");
const tg_file_id_1 = require("tg-file-id");
const Medias = __importStar(require("../../Utils/Medias"));
const big_integer_1 = __importDefault(require("big-integer"));
const GetFileInfo_1 = require("./GetFileInfo");
const path_1 = __importDefault(require("path"));
const Error_1 = __importDefault(require("../../Context/Error"));
function clean(more) {
    if (more) {
        let purge = [
            'workers',
            'mimeType',
            'duration',
            'onProgress',
            'download',
            'thumbnail',
            'performer',
            'title',
            'filename',
        ];
        for (let [key] of Object.entries(more)) {
            if (purge.includes(key))
                delete more[key];
        }
    }
    return more;
}
/**
 * Sending Voice
 * @param snakeClient - Client
 * @param {string|number|bigint} chatId - Chat/Groups/Channel id.
 * @param {string|Buffer} fileId - File Location/Url/Buffer .
 * @param {Object} more - more parameters to use.
 * ```ts
 * bot.command("voice",async (ctx) => {
 *     let results = await ctx.telegram.sendVoice(ctx.chat.id,"file id here")
 * })
 * ```
 */
async function SendVoice(snakeClient, chatId, fileId, more) {
    var _a, _b, _c, _d, _e, _f, _g;
    try {
        snakeClient.log.debug('Running telegram.sendVoice');
        if (typeof chatId === 'number')
            snakeClient.log.warning('Type of chatId is number, please switch to BigInt or String for security Ids 64 bit int.');
        const validateThumb = (thumb) => {
            switch (thumb.constructor) {
                case Buffer:
                    return true;
                    break;
                case String:
                    thumb;
                    if (/^http/i.test(String(thumb)) || /^(\/|\.\.?\/|~\/)/i.test(String(thumb)))
                        return true;
                    return false;
                    break;
                default:
                    return false;
            }
        };
        switch (fileId.constructor) {
            case Buffer:
                fileId;
                let bufferInfo = await (0, GetFileInfo_1.GetFileInfo)(fileId);
                return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaUploadedDocument(Object.assign({ 
                    //@ts-ignore
                    file: await (0, UploadFile_1.UploadFile)(snakeClient, bufferInfo.source, {
                        workers: (more === null || more === void 0 ? void 0 : more.workers) || 1,
                        onProgress: more === null || more === void 0 ? void 0 : more.onProgress,
                        fileName: more === null || more === void 0 ? void 0 : more.filename,
                    }), 
                    //@ts-ignore
                    mimeType: bufferInfo.mime || (more === null || more === void 0 ? void 0 : more.mimeType) || 'audio/ogg', attributes: [
                        new telegram_1.Api.DocumentAttributeAudio({
                            duration: (_a = more === null || more === void 0 ? void 0 : more.duration) !== null && _a !== void 0 ? _a : 0,
                            voice: true,
                            title: (_b = more === null || more === void 0 ? void 0 : more.title) !== null && _b !== void 0 ? _b : 'unknown',
                            performer: (_c = more === null || more === void 0 ? void 0 : more.performer) !== null && _c !== void 0 ? _c : 'unknown',
                        }),
                        new telegram_1.Api.DocumentAttributeFilename({
                            fileName: (more === null || more === void 0 ? void 0 : more.filename) || 'unknown',
                        }),
                    ], forceFile: false }, (more && more.thumbnail && validateThumb(more.thumbnail)
                    ? {
                        thumb: await (0, UploadFile_1.UploadFile)(snakeClient, more.thumbnail, {
                            workers: more.workers || 1,
                        }),
                    }
                    : {}))), clean(more));
                break;
            case Medias.MediaVoice:
                return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaDocument({
                    id: new telegram_1.Api.InputDocument({
                        //@ts-ignore
                        id: (0, big_integer_1.default)(String(fileId._id)),
                        //@ts-ignore
                        accessHash: (0, big_integer_1.default)(String(fileId._accessHash)),
                        //@ts-ignore
                        fileReference: Buffer.from(fileId._fileReference, 'hex'),
                    }),
                }));
                break;
            case String:
                fileId;
                if (/^http/i.test(String(fileId)) || /^(\/|\.\.?\/|~\/)/i.test(String(fileId))) {
                    let download = (_d = more === null || more === void 0 ? void 0 : more.download) !== null && _d !== void 0 ? _d : false;
                    if (/^http/i.test(String(fileId)) && !download) {
                        let file = new telegram_1.Api.InputMediaDocumentExternal({
                            url: fileId,
                        });
                        return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, file, clean(more));
                    }
                    let fileInfo = await (0, GetFileInfo_1.GetFileInfo)(String(fileId));
                    let basename = path_1.default.basename(String(fileId));
                    if (!/\.([0-9a-z]+)(?=[?#])|(\.)(?:[\w]+)$/gim.exec(basename)) {
                        basename = `${basename}.${fileInfo === null || fileInfo === void 0 ? void 0 : fileInfo.ext}`;
                    }
                    return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaUploadedDocument(Object.assign({ 
                        //@ts-ignore
                        file: await (0, UploadFile_1.UploadFile)(snakeClient, fileInfo.source, {
                            workers: (more === null || more === void 0 ? void 0 : more.workers) || 1,
                            fileName: basename,
                            onProgress: more === null || more === void 0 ? void 0 : more.onProgress,
                        }), 
                        //@ts-ignore
                        mimeType: fileInfo.mime || (more === null || more === void 0 ? void 0 : more.mimeType) || 'audio/ogg', attributes: [
                            new telegram_1.Api.DocumentAttributeAudio({
                                duration: (_e = more === null || more === void 0 ? void 0 : more.duration) !== null && _e !== void 0 ? _e : 0,
                                voice: true,
                                title: (_f = more === null || more === void 0 ? void 0 : more.title) !== null && _f !== void 0 ? _f : 'unknown',
                                performer: (_g = more === null || more === void 0 ? void 0 : more.performer) !== null && _g !== void 0 ? _g : 'unknown',
                            }),
                            new telegram_1.Api.DocumentAttributeFilename({
                                fileName: (more === null || more === void 0 ? void 0 : more.filename) || basename || 'unknown',
                            }),
                        ], forceFile: false }, (more && more.thumbnail && validateThumb(more.thumbnail)
                        ? {
                            thumb: await (0, UploadFile_1.UploadFile)(snakeClient, more.thumbnail, {
                                workers: more.workers || 1,
                            }),
                        }
                        : {}))), clean(more));
                }
                else {
                    //@ts-ignore
                    let file = await (0, tg_file_id_1.decodeFileId)(fileId);
                    if (file.typeId !== 4)
                        throw new Error(`Invalid fileId. This fileId not for voice`);
                    return await (0, SendMedia_1.SendMedia)(snakeClient, chatId, new telegram_1.Api.InputMediaDocument({
                        id: new telegram_1.Api.InputDocument({
                            id: (0, big_integer_1.default)(String(file.id)),
                            accessHash: (0, big_integer_1.default)(String(file.access_hash)),
                            fileReference: Buffer.from(file.fileReference, 'hex'),
                        }),
                    }));
                }
                break;
            default:
                throw new Error(`Couldn't resolve this fileId.`);
        }
    }
    catch (error) {
        snakeClient.log.error('Failed to running telegram.sendVoice');
        throw new Error_1.default(error.message, 'telegram.sendVoice', `${chatId},${Buffer.isBuffer(fileId) ? `<Buffer ${fileId.toString('hex')}>` : JSON.stringify(fileId)}${more ? ',' + JSON.stringify(more) : ''}`);
    }
}
exports.SendVoice = SendVoice;
