// tgsnake/lib/Update/UpdateBotShippingQuery.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://github.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateBotShippingQuery = void 0;
const From_1 = require("../Utils/From");
const Update_1 = require("./Update");
const Payment_1 = require("../Utils/Payment");
class UpdateBotShippingQuery extends Update_1.Update {
    constructor() {
        super();
        this['_'] = 'updateBotShippingQuery';
    }
    async init(update, SnakeClient) {
        SnakeClient.log.debug(`Creating ${this['_']}`);
        this.telegram = SnakeClient.telegram;
        this.id = BigInt(String(update.queryId));
        this.payload = update.payload.toString('utf8');
        this.shippingAddress = new Payment_1.PostAddress(update.shippingAddress);
        this.from = new From_1.From();
        await this.from.init(BigInt(String(update.userId)), SnakeClient);
        return this;
    }
    async answer(options, error) {
        return this.telegram.answerShippingQuery(this.id, options, error);
    }
}
exports.UpdateBotShippingQuery = UpdateBotShippingQuery;
