// tgsnake/lib/Utils/Medias/Animated.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaAnimation = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const telegram_1 = require("telegram");
const Media_1 = require("./Media");
const tg_file_id_1 = require("tg-file-id");
const big_integer_1 = __importDefault(require("big-integer"));
const CleanObject_1 = require("../CleanObject");
const Error_1 = __importDefault(require("../../Context/Error"));
class MediaAnimation extends Media_1.Media {
    constructor() {
        super();
        this['_'] = 'animation';
    }
    async encode(animated, snakeClient) {
        var _a;
        snakeClient.log.debug('Creating MediaAnimation');
        this.snakeClient = snakeClient;
        this._id = BigInt(String(animated.id));
        this._accessHash = BigInt(String(animated.accessHash));
        this._fileReference = animated.fileReference.toString('hex');
        this.mimeType = animated.mimeType;
        this.size = BigInt(String(animated.size));
        this.dcId = animated.dcId;
        for (let attribute of animated.attributes) {
            if (attribute instanceof telegram_1.Api.DocumentAttributeVideo) {
                attribute;
                this.supportStreaming = (_a = attribute.supportsStreaming) !== null && _a !== void 0 ? _a : false;
                this.duration = attribute.duration;
                this.width = attribute.w;
                this.height = attribute.h;
            }
            if (attribute instanceof telegram_1.Api.DocumentAttributeFilename) {
                attribute;
                this.filename = attribute.fileName;
            }
            if (attribute instanceof telegram_1.Api.DocumentAttributeHasStickers) {
                attribute;
                this.hasSticker = true;
            }
        }
        let file = new tg_file_id_1.FileId();
        file.version = 4;
        file.subVersion = 30;
        file.dcId = this.dcId;
        file.fileType = 'animation';
        file.typeId = 10;
        file.volumeId = BigInt(1);
        file.id = this._id;
        file.accessHash = this._accessHash;
        file.fileReference = this._fileReference;
        this.fileId = await file.toFileId();
        this.uniqueFileId = await file.toFileUniqId();
        await (0, CleanObject_1.Cleaning)(this);
        return this;
    }
    async download(fileId, params) {
        this.snakeClient.log.debug('Downloading animation');
        const { client, log } = this.snakeClient;
        const file = fileId !== null && fileId !== void 0 ? fileId : this.fileId;
        if (!file) {
            this.snakeClient.log.error('Failed to download animation cause: FileId not found!');
            throw new Error_1.default('FileId not found!', 'Animation.download', String(file));
        }
        if (file !== this.fileId) {
            let dFile = this.decode(file);
            if (dFile.typeId !== 4) {
                this.snakeClient.log.error('Failed to download animation cause: Miss match file type!');
                throw new Error_1.default('Miss match file type!', 'Animation.download', String(file));
            }
            let dParams = Object.assign({
                dcId: dFile.dcId,
                progressCallback: (progress) => {
                    return log.debug(`Downloading animation [${Math.round(progress)}]`);
                },
            }, params !== null && params !== void 0 ? params : {});
            if (dParams.fileSize && typeof dParams.fileSize == 'bigint') {
                // @ts-ignore
                dParams.fileSize = (0, big_integer_1.default)(String(dParams.fileSize));
            }
            return client.downloadFile(new telegram_1.Api.InputDocumentFileLocation({
                id: (0, big_integer_1.default)(String(dFile.id)),
                accessHash: (0, big_integer_1.default)(String(dFile.access_hash)),
                fileReference: Buffer.from(dFile.fileReference, 'hex'),
                thumbSize: 'w',
            }), 
            // @ts-ignore
            dParams);
        }
        let dParams = Object.assign({
            dcId: this.dcId,
            fileSize: (0, big_integer_1.default)(this.size),
            progressCallback: (progress) => {
                return log.debug(`Downloading animation [${Math.round(progress)}]`);
            },
        }, params !== null && params !== void 0 ? params : {});
        if (dParams.fileSize && typeof dParams.fileSize == 'bigint') {
            // @ts-ignore
            dParams.fileSize = (0, big_integer_1.default)(String(dParams.fileSize));
        }
        return client.downloadFile(new telegram_1.Api.InputDocumentFileLocation({
            id: (0, big_integer_1.default)(String(this._id)),
            accessHash: (0, big_integer_1.default)(String(this._accessHash)),
            fileReference: Buffer.from(this._fileReference, 'hex'),
            thumbSize: 'w',
        }), 
        // @ts-ignore
        dParams);
    }
}
exports.MediaAnimation = MediaAnimation;
