// tgsnake/lib/Utils/Medias/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateMedia = exports.MediaChatPhoto = exports.MediaWebPage = exports.MediaInvoice = exports.MediaGame = exports.MediaVoice = exports.MediaAudio = exports.MediaDocument = exports.MediaAnimation = exports.MediaVideo = exports.MediaVideoNote = exports.MediaVenue = exports.MediaPhoto = exports.MediaLocation = exports.MediaDice = exports.MediaPoll = exports.MediaSticker = exports.MediaContact = exports.Media = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
var Media_1 = require("./Media");
Object.defineProperty(exports, "Media", { enumerable: true, get: function () { return Media_1.Media; } });
var Contact_1 = require("./Contact");
Object.defineProperty(exports, "MediaContact", { enumerable: true, get: function () { return Contact_1.MediaContact; } });
var Sticker_1 = require("./Sticker");
Object.defineProperty(exports, "MediaSticker", { enumerable: true, get: function () { return Sticker_1.MediaSticker; } });
var Poll_1 = require("./Poll");
Object.defineProperty(exports, "MediaPoll", { enumerable: true, get: function () { return Poll_1.MediaPoll; } });
var Dice_1 = require("./Dice");
Object.defineProperty(exports, "MediaDice", { enumerable: true, get: function () { return Dice_1.MediaDice; } });
var Location_1 = require("./Location");
Object.defineProperty(exports, "MediaLocation", { enumerable: true, get: function () { return Location_1.MediaLocation; } });
var Photo_1 = require("./Photo");
Object.defineProperty(exports, "MediaPhoto", { enumerable: true, get: function () { return Photo_1.MediaPhoto; } });
var Venue_1 = require("./Venue");
Object.defineProperty(exports, "MediaVenue", { enumerable: true, get: function () { return Venue_1.MediaVenue; } });
var VideoNote_1 = require("./VideoNote");
Object.defineProperty(exports, "MediaVideoNote", { enumerable: true, get: function () { return VideoNote_1.MediaVideoNote; } });
var Video_1 = require("./Video");
Object.defineProperty(exports, "MediaVideo", { enumerable: true, get: function () { return Video_1.MediaVideo; } });
var Animated_1 = require("./Animated");
Object.defineProperty(exports, "MediaAnimation", { enumerable: true, get: function () { return Animated_1.MediaAnimation; } });
var Document_1 = require("./Document");
Object.defineProperty(exports, "MediaDocument", { enumerable: true, get: function () { return Document_1.MediaDocument; } });
var Audio_1 = require("./Audio");
Object.defineProperty(exports, "MediaAudio", { enumerable: true, get: function () { return Audio_1.MediaAudio; } });
var Voice_1 = require("./Voice");
Object.defineProperty(exports, "MediaVoice", { enumerable: true, get: function () { return Voice_1.MediaVoice; } });
var Game_1 = require("./Game");
Object.defineProperty(exports, "MediaGame", { enumerable: true, get: function () { return Game_1.MediaGame; } });
var Invoice_1 = require("./Invoice");
Object.defineProperty(exports, "MediaInvoice", { enumerable: true, get: function () { return Invoice_1.MediaInvoice; } });
var WebPage_1 = require("./WebPage");
Object.defineProperty(exports, "MediaWebPage", { enumerable: true, get: function () { return WebPage_1.MediaWebPage; } });
var ChatPhoto_1 = require("./ChatPhoto");
Object.defineProperty(exports, "MediaChatPhoto", { enumerable: true, get: function () { return ChatPhoto_1.MediaChatPhoto; } });
var Generator_1 = require("./Generator");
Object.defineProperty(exports, "GenerateMedia", { enumerable: true, get: function () { return Generator_1.GenerateMedia; } });
