// tgsnake/lib/Utils/Medias/Location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaLocation = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const telegram_1 = require("telegram");
const Media_1 = require("./Media");
const CleanObject_1 = require("../CleanObject");
class MediaLocation extends Media_1.Media {
    constructor() {
        super();
        this['_'] = 'location';
    }
    async encode(location, snakeClient) {
        snakeClient.log.debug('Creating MediaLocation');
        this.snakeClient = snakeClient;
        const generate = async (geo) => {
            var _a, _b;
            this.latitude = geo.lat;
            this.longitude = geo.long;
            this.accessHash = BigInt(String((_a = geo.accessHash) !== null && _a !== void 0 ? _a : 0));
            this.accuracyRadius = (_b = geo.accuracyRadius) !== null && _b !== void 0 ? _b : 0;
            await (0, CleanObject_1.Cleaning)(this);
            return this;
        };
        if (location instanceof telegram_1.Api.MessageMediaGeo) {
            location;
            if (location.geo instanceof telegram_1.Api.GeoPoint) {
                location.geo;
                return generate(location.geo);
            }
        }
        if (location instanceof telegram_1.Api.GeoPoint) {
            location;
            return generate(location);
        }
        return this;
    }
}
exports.MediaLocation = MediaLocation;
