// tgsnake/lib/Utils/Medias/Sticker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaSticker = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const telegram_1 = require("telegram");
const Media_1 = require("./Media");
const tg_file_id_1 = require("tg-file-id");
const big_integer_1 = __importDefault(require("big-integer"));
const CleanObject_1 = require("../CleanObject");
const Error_1 = __importDefault(require("../../Context/Error"));
class MediaSticker extends Media_1.Media {
    constructor() {
        super();
        this['_'] = 'sticker';
    }
    async encode(sticker, snakeClient) {
        var _a, _b, _c, _d;
        snakeClient.log.debug('Creating MediaSticker');
        this.snakeClient = snakeClient;
        this.mimeType = sticker.mimeType;
        this.dcId = sticker.dcId;
        this.isAnimated = sticker.mimeType === 'application/x-tgsticker';
        this.isVideo = sticker.mimeType === 'video/webm';
        this.width = 512;
        this.height = 512;
        this.size = BigInt(String(sticker.size));
        this._id = BigInt(String(sticker.id));
        this._accessHash = BigInt(String(sticker.accessHash));
        this._stickerSetId = BigInt(0);
        this._stickerSetAccessHash = BigInt(0);
        this._fileReference = sticker.fileReference.toString('hex');
        let file = new tg_file_id_1.FileId();
        file.version = 4;
        file.subVersion = 30;
        file.dcId = sticker.dcId;
        file.fileType = 'sticker';
        file.typeId = 8;
        file.volumeId = BigInt(1);
        file.id = BigInt(String(sticker.id));
        file.accessHash = BigInt(String(sticker.accessHash));
        file.fileReference = sticker.fileReference.toString('hex');
        for (let attributes of sticker.attributes) {
            if (attributes instanceof telegram_1.Api.DocumentAttributeSticker) {
                attributes;
                this.emoji = attributes.alt;
                if (attributes.stickerset instanceof telegram_1.Api.InputStickerSetID) {
                    attributes.stickerset;
                    this._stickerSetId = BigInt(String((_a = attributes.stickerset.id) !== null && _a !== void 0 ? _a : 0));
                    this._stickerSetAccessHash = BigInt(String((_b = attributes.stickerset.accessHash) !== null && _b !== void 0 ? _b : 0));
                }
                if (attributes.stickerset instanceof telegram_1.Api.InputStickerSetShortName) {
                    attributes.stickerset;
                    this._stickerSetShortName = attributes.stickerset.shortName;
                }
            }
            if (attributes instanceof telegram_1.Api.DocumentAttributeImageSize) {
                attributes;
                this.width = (_c = attributes.w) !== null && _c !== void 0 ? _c : 512;
                this.height = (_d = attributes.h) !== null && _d !== void 0 ? _d : 512;
            }
        }
        file.stickerSetId = this._stickerSetId;
        file.stickerSetAccessHash = this._stickerSetAccessHash;
        this.fileId = await file.toFileId();
        this.uniqueFileId = await file.toFileUniqId();
        await (0, CleanObject_1.Cleaning)(this);
        return this;
    }
    async download(fileId, params) {
        this.snakeClient.log.debug('Downloading sticker');
        const { client, log } = this.snakeClient;
        const file = fileId !== null && fileId !== void 0 ? fileId : this.fileId;
        if (!file) {
            this.snakeClient.log.error('Failed to download sticker cause: FileId not found!');
            throw new Error_1.default('FileId not found!', 'Sticker.download', String(file));
        }
        if (file !== this.fileId) {
            let dFile = this.decode(file);
            if (dFile.typeId !== 8) {
                this.snakeClient.log.error('Failed to download sticker cause: Miss match file type!');
                throw new Error_1.default('Miss match file type!', 'Sticker.download', String(file));
            }
            let dParams = Object.assign({
                dcId: dFile.dcId,
                progressCallback: (progress) => {
                    return log.debug(`Downloading Sticker [${Math.round(progress)} %]`);
                },
            }, params !== null && params !== void 0 ? params : {});
            if (dParams.fileSize && typeof dParams.fileSize == 'bigint') {
                // @ts-ignore
                dParams.fileSize = (0, big_integer_1.default)(String(dParams.fileSize));
            }
            return client.downloadFile(new telegram_1.Api.InputDocumentFileLocation({
                id: (0, big_integer_1.default)(String(dFile.id)),
                accessHash: (0, big_integer_1.default)(String(dFile.access_hash)),
                fileReference: Buffer.from(dFile.fileReference, 'hex'),
                thumbSize: 'w',
            }), 
            // @ts-ignore
            dParams);
        }
        let dParams = Object.assign({
            dcId: this.dcId,
            fileSize: (0, big_integer_1.default)(this.size),
            progressCallback: (progress) => {
                return log.debug(`Downloading Sticker [${Math.round(progress)}]`);
            },
        }, params !== null && params !== void 0 ? params : {});
        if (dParams.fileSize && typeof dParams.fileSize == 'bigint') {
            // @ts-ignore
            dParams.fileSize = (0, big_integer_1.default)(String(dParams.fileSize));
        }
        return client.downloadFile(new telegram_1.Api.InputDocumentFileLocation({
            id: (0, big_integer_1.default)(String(this._id)),
            accessHash: (0, big_integer_1.default)(String(this._accessHash)),
            fileReference: Buffer.from(this._fileReference, 'hex'),
            thumbSize: 'w',
        }), 
        // @ts-ignore
        dParams);
    }
}
exports.MediaSticker = MediaSticker;
