// tgsnake/lib/Utils/Medias/Venue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaVenue = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const telegram_1 = require("telegram");
const Media_1 = require("./Media");
const CleanObject_1 = require("../CleanObject");
class MediaVenue extends Media_1.Media {
    constructor() {
        super();
        this['_'] = 'venue';
    }
    async encode(venue, snakeClient) {
        var _a, _b;
        snakeClient.log.debug('Creating MediaVenue');
        this.snakeClient = snakeClient;
        if (venue.geo instanceof telegram_1.Api.GeoPoint) {
            const geo = venue.geo;
            this.latitude = geo.lat;
            this.longitude = geo.long;
            this.accessHash = BigInt(String((_a = geo.accessHash) !== null && _a !== void 0 ? _a : 0));
            this.accuracyRadius = (_b = geo.accuracyRadius) !== null && _b !== void 0 ? _b : 0;
        }
        this.title = venue.title;
        this.provider = venue.provider;
        this.address = venue.address;
        this.id = venue.venueId;
        this.type = venue.venueType;
        await (0, CleanObject_1.Cleaning)(this);
        return this;
    }
}
exports.MediaVenue = MediaVenue;
