// tgsnake/lib/Utils/Medias/WebPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaWebPage = void 0;
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
const telegram_1 = require("telegram");
const Media_1 = require("./Media");
const Photo_1 = require("./Photo");
// TODO:
// convert field document,cachedPage to tgsnake json
class MediaWebPage extends Media_1.Media {
    constructor() {
        super();
        this['_'] = 'webPage';
    }
    async encode(webPage, snakeClient) {
        snakeClient.log.debug('Creating MediaWebPage');
        this.snakeClient = snakeClient;
        //@ts-ignore
        if (webPage.id !== undefined)
            this.id = BigInt(String(webPage.id));
        if (webPage instanceof telegram_1.Api.WebPage) {
            webPage;
            this.url = webPage.url;
            this.displayUrl = webPage.displayUrl;
            this.hash = webPage.hash;
            this.type = webPage.type;
            this.siteName = webPage.siteName;
            this.title = webPage.title;
            this.description = webPage.description;
            this.embedUrl = webPage.embedUrl;
            this.embedType = webPage.embedType;
            this.embedWidth = webPage.embedWidth;
            this.embedHeight = webPage.embedHeight;
            this.duration = webPage.duration;
            this.author = webPage.author;
            if (webPage.photo) {
                this.photo = new Photo_1.MediaPhoto();
                await this.photo.encode(webPage.photo, snakeClient);
            }
        }
        return this;
    }
}
exports.MediaWebPage = MediaWebPage;
