// tgsnake/lib/Utils/ForwardMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForwardMessage = void 0;
const Chat_1 = require("./Chat");
const From_1 = require("./From");
const CleanObject_1 = require("./CleanObject");
class ForwardMessage {
    constructor() { }
    async init(forwardHeader, snakeClient) {
        snakeClient.log.debug(`Creating forwardMessage`);
        this.imported = forwardHeader.imported;
        this.fromName = forwardHeader.fromName;
        this.date = forwardHeader.date;
        this.channelPost = forwardHeader.channelPost;
        this.postAuthor = forwardHeader.postAuthor;
        this.savedFromMsgId = forwardHeader.savedFromMsgId;
        this.psaType = forwardHeader.psaType;
        if (forwardHeader.fromId) {
            try {
                let from = new From_1.From();
                await from.init(forwardHeader.fromId, snakeClient);
                this.from = from;
            }
            catch (error) { }
        }
        if (forwardHeader.savedFromPeer) {
            try {
                let chat = new Chat_1.Chat();
                await chat.init(forwardHeader.savedFromPeer, snakeClient);
                this.savedFromChat = chat;
            }
            catch (error) { }
        }
        await (0, CleanObject_1.Cleaning)(this);
        return this;
    }
}
exports.ForwardMessage = ForwardMessage;
