// tgsnake/lib/Utils/From.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
Object.defineProperty(exports, "__esModule", { value: true });
exports.From = void 0;
const telegram_1 = require("telegram");
const ToBigInt_1 = require("./ToBigInt");
const big_integer_1 = require("big-integer");
const CleanObject_1 = require("./CleanObject");
class From {
    constructor() { }
    async init(peer, snakeClient) {
        let id;
        if (typeof peer !== 'number' && typeof peer !== 'bigint' && typeof peer !== 'string') {
            if (peer instanceof telegram_1.Api.PeerUser) {
                peer;
                if ((0, big_integer_1.isInstance)(peer.userId)) {
                    //@ts-ignore
                    id = BigInt((0, ToBigInt_1.toString)(peer.userId));
                }
                else {
                    //@ts-ignore
                    id = BigInt(peer.userId);
                }
            }
            if (peer instanceof telegram_1.Api.PeerChat) {
                peer;
                if ((0, big_integer_1.isInstance)(peer.chatId)) {
                    //@ts-ignore
                    id = BigInt(Number(`-${(0, ToBigInt_1.toString)(peer.chatId)}`));
                }
                else {
                    //@ts-ignore
                    id = BigInt(Number(`-${peer.chatId}`));
                }
            }
            if (peer instanceof telegram_1.Api.PeerChannel) {
                peer;
                if ((0, big_integer_1.isInstance)(peer.channelId)) {
                    //@ts-ignore
                    id = BigInt(Number(`-100${(0, ToBigInt_1.toString)(peer.channelId)}`));
                }
                else {
                    //@ts-ignore
                    id = BigInt(Number(`-100${peer.channelId}`));
                }
            }
        }
        else {
            id = typeof peer == 'number' ? BigInt(peer) : peer;
        }
        if (id) {
            snakeClient.log.debug(`Creating User ${id}`);
            try {
                let entity = await snakeClient.telegram.getEntity(id, true);
                this.id = entity.id;
                this.username = entity.username;
                this.firstName = entity.firstName;
                this.lastName = entity.lastName;
                this.status = entity.status;
                this.self = entity.self;
                this.deleted = entity.deleted;
                this.fake = entity.fake;
                this.scam = entity.scam;
                this.bot = entity.bot;
                this.verified = entity.verified;
                this.restricted = entity.restricted;
                this.dcId = entity.dcId;
                this.photo = entity.photo;
                this.restrictionReason = entity.restrictionReason;
                this.accessHash = entity.accessHash;
            }
            catch (error) {
                return this;
            }
        }
        await (0, CleanObject_1.Cleaning)(this);
        return this;
    }
    /** @hidden */
    toJSON() {
        let obj = (0, CleanObject_1.betterConsoleLog)(this);
        for (let [key, value] of Object.entries(obj)) {
            if (typeof value == 'bigint')
                obj[key] = String(value);
        }
        return obj;
    }
}
exports.From = From;
