// tgsnake/lib/Utils/Payment.js
"use strict";
// Tgsnake - Telegram MTProto framework developed based on gram.js.
// Copyright (C) 2022 Butthx <https://guthub.com/butthx>
//
// This file is part of Tgsnake
//
// Tgsnake is a free software : you can redistribute it and/or modify
//  it under the terms of the MIT License as published.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentCharge = exports.PostAddress = exports.PaymentRequestedInfo = void 0;
class PaymentRequestedInfo {
    constructor(paymentRequest) {
        if (paymentRequest.name)
            this.name = paymentRequest.name;
        if (paymentRequest.phone)
            this.phone = paymentRequest.phone;
        if (paymentRequest.email)
            this.email = paymentRequest.email;
        if (paymentRequest.shippingAddress)
            this.shippingAddress = new PostAddress(paymentRequest.shippingAddress);
    }
}
exports.PaymentRequestedInfo = PaymentRequestedInfo;
class PostAddress {
    constructor(postAddress) {
        this.streetLine1 = postAddress.streetLine1;
        this.streetLine2 = postAddress.streetLine2;
        this.city = postAddress.city;
        this.state = postAddress.state;
        this.countryIso2 = postAddress.countryIso2;
        this.postCode = postAddress.postCode;
        return this;
    }
}
exports.PostAddress = PostAddress;
class PaymentCharge {
    constructor(paymentCharge) {
        this.id = paymentCharge.id;
        this.providerChargeId = paymentCharge.providerChargeId;
        return this;
    }
}
exports.PaymentCharge = PaymentCharge;
